pkg/sensors/accelerator/gpu/source/gpu_dcgm.go

```diff
@@ -0,0 +1,389 @@
+//go:build gpu
+// +build gpu
+
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/NVIDIA/go-dcgm/pkg/dcgm"
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"github.com/sustainable-computing-io/kepler/pkg/config"
+	"k8s.io/klog/v2"
+)
+
+const (
+	debugLevel = 5
+)
+
+var (
+	deviceFields []dcgm.Short = []dcgm.Short{
+		// https://docs.nvidia.com/datacenter/dcgm/1.7/dcgm-api/group__dcgmFieldIdentifiers.htm
+		dcgm.DCGM_FI_PROF_PIPE_TENSOR_ACTIVE,
+	}
+	deviceFieldsString = []string{
+		"dcgm.DCGM_FI_PROF_PIPE_TENSOR_ACTIVE",
+	}
+	ratioFields              uint = dcgm.DCGM_FI_PROF_PIPE_TENSOR_ACTIVE // this is the field that we will use to calculate the utilization per @yuezhu1
+	SkipDCGMValue                 = "SKIPPING DCGM VALUE"
+	FailedToConvert               = "ERROR - FAILED TO CONVERT TO STRING"
+	gpuMigArray              [][]MigDevice
+	totalMultiProcessorCount map[string]int
+)
+
+type GPUDcgm struct {
+	collectionSupported bool
+	libInited           bool
+	devices             map[string]interface{}
+	deviceGroupName     string
+	deviceGroupHandle   dcgm.GroupHandle
+	fieldGroupName      string
+	fieldGroupHandle    dcgm.FieldHandle
+	pidGroupName        string
+	pidGroupHandle      dcgm.GroupHandle // TODO: wait till https://github.com/NVIDIA/go-dcgm/issues/59 is resolved
+	entities            map[string]dcgm.GroupEntityPair
+	cleanup             func()
+}
+
+func (d *GPUDcgm) GetName() string {
+	return "dcgm"
+}
+
+func (d *GPUDcgm) InitLib() error {
+	d.devices = make(map[string]interface{})
+	d.entities = make(map[string]dcgm.GroupEntityPair)
+
+	// cleanup, err := dcgm.Init(dcgm.Embedded) // embeded mode is not recommended for production per https://github.com/NVIDIA/dcgm-exporter/issues/22#issuecomment-1321521995
+	cleanup, err := dcgm.Init(dcgm.Standalone, config.DCGMHostEngineEndpoint, "0")
```

Comment:
this requires :
- the host must enable dcgm service and run nv-hostengine
- pod spec to enable `hostNetwork` so to access the nvidia dcgm hostengine

Reply:
To investigate how to use `embeded` mode after this PR is merged.

Reply:
I forgot to mention that, the `standalone` connecting to the localhost works in K8s environments.

Reply:
ok, let's consider it as an enhancement then.

Reply:
here is the issue https://github.com/sustainable-computing-io/kepler/issues/1243